### PR TITLE
Aggregation revamp.

### PR DESCRIPTION
**Pull request by ryzhyk**
See commit messages for details.

@mbudiu-vmw , this is still groupby-aggregate only.  I will add an aggregation operator that works on the entire Z-set shortly.

Some other TODOs:

- Proptests
- Optimized Max and Top-k aggregates.
- Argmin, Argmax and other useful library aggregates (@mbudiu-vmw , what do we need for SQL?)

### CODE REVIEW

File: src/operator/aggregate/fold.rs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+/// applying a user-provided `output` function to the final value of
+/// the accumulator.
+///
+/// # Type arguments
+///
+/// * `A` - accumulator
+/// * `SF` - step function
+/// * `OF` - output function
+#[derive(Clone)]
+pub struct Fold<A, SF, OF> {
```
Comment (mbudiu-vmw) [inline src/operator/aggregate/fold.rs:19]: Call this FoldOps?

File: src/operator/aggregate/fold.rs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+                (self.step)(&mut acc, cursor.key(), weight);
+            }
+
+            cursor.step_key();
+        }
+
+        if non_empty {
+            Some((self.output)(acc))
+        } else {
+            None
```
Comment (mbudiu-vmw) [inline src/operator/aggregate/fold.rs:83]: This needs to be documented.

File: src/operator/aggregate/fold.rs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    {
+        let mut acc = self.init.clone();
+        let mut non_empty = false;
+
+        while cursor.key_valid() {
+            let mut weight = R::zero();
+
+            cursor.map_times(|_t, w| {
+                weight.add_assign_by_ref(w);
+            });
```
Comment (mbudiu-vmw) [inline src/operator/aggregate/fold.rs:71]: This is asymmetric, it doesn't accumulate elements with zero weight at the start but it does afterwards?

File: src/operator/aggregate/fold.rs [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    pub fn new(init: A, step: SF) -> Self {
+        Self {
+            init,
+            step,
+            output: identity,
+        }
+    }
+}
+
+impl<A, SF, OF> Fold<A, SF, OF> {
```
Comment (mbudiu-vmw) [inline src/operator/aggregate/fold.rs:41]: How about a parallel composition of fold which makes a fold on tuples out of n folds point wise? May need to be a macro.

File: src/operator/aggregate/mod.rs [hunk 5]
```diff
@@ -0,0 +1,1058 @@
...
+/// aggregators.  Users will typicaly work with ready-made implementations
+/// like [`Min`] and [`Fold`].
+pub trait Aggregator<K, T, R> {
+    /// Aggregate type output by this aggregator.
+    type Output;
+
+    /// Compute an aggregate over a Z-set.
+    ///
+    /// Takes a cursor pointing to the first key of a Z-set and outputs
+    /// an aggregate of the Z-set.  This is a low-level API that
```
Comment (mbudiu-vmw) [inline src/operator/aggregate/mod.rs:58]: Duplicated below

File: src/operator/aggregate/mod.rs [hunk 6]
```diff
@@ -0,0 +1,1058 @@
...
+struct WeightAggregator;
+
+impl<T, R> Aggregator<(), T, R> for WeightAggregator
+where
+    T: Timestamp,
+    R: MonoidValue,
+{
+    type Output = R;
+
+    fn aggregate<'s, C>(&mut self, cursor: &mut C) -> Option<Self::Output>
```
Comment (mbudiu-vmw) [inline src/operator/aggregate/mod.rs:83]: This is more like weighted count.

File: src/operator/aggregate/mod.rs [hunk 7]
```diff
@@ -0,0 +1,1058 @@
...
+    ///
+    /// ```text
+    ///    __
+    ///    ╲
+    ///    ╱ f(k,v) * w
+    ///    ‾‾
+    /// (k,v,w) ∈ Z
+    /// ```
+    ///
+    /// This is a linear operator.
```
Comment (mbudiu-vmw) [inline src/operator/aggregate/mod.rs:283]: Should this be a property?

File: src/operator/aggregate/mod.rs [hunk 8]
```diff
@@ -0,0 +1,1058 @@
...
+            time: IT::Time::clock_start(),
+            empty_input: false,
+            empty_output: false,
+            keys_of_interest: BTreeMap::new(),
+            output_delta: Vec::new(),
+            _type: PhantomData,
+        }
+    }
+
+    /// Compute output of the operator for `key`.
```
Comment (mbudiu-vmw) [inline src/operator/aggregate/mod.rs:456]: Can't say I got all this. Looking forward to the paper.

**Review by mbudiu-vmw — APPROVED** [review]
SQL logic test doesn't use other aggregates. Some dialects of SQL may have more. Aggregate distinct will în fact have to call distinct and then a normal aggregate.

File: src/operator/aggregate/fold.rs (hunk 3 above)
Comment (ryzhyk) [reply src/operator/aggregate/fold.rs:71]: It should always skip zero-weight elements in the cursor. It handles all keys in the cursor in the same way.

File: src/operator/aggregate/mod.rs (hunk 7 above)
Comment (ryzhyk) [reply src/operator/aggregate/mod.rs:283]: What do you mean by this?

File: src/operator/aggregate/fold.rs (hunk 3 above)
Comment (mbudiu-vmw) [reply src/operator/aggregate/fold.rs:71]: Doesn't seem to do that. E.g. If the weight becomes zero in the middle.

File: src/operator/aggregate/fold.rs (hunk 4 above)
Comment (ryzhyk) [reply src/operator/aggregate/fold.rs:41]: Yep, that make sense. I'll create an issue (need to work on something else now).

File: src/operator/aggregate/fold.rs (hunk 3 above)
Comment (ryzhyk) [reply src/operator/aggregate/fold.rs:71]: It will continue iterating over other keys in the same way.

File: src/operator/aggregate/fold.rs [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+    {
+        let mut acc = self.init.clone();
+        let mut non_empty = false;
+
+        while cursor.key_valid() {
+            let mut weight = R::zero();
+
+            cursor.map_times(|_t, w| {
+                weight.add_assign_by_ref(w);
+            });
```
Comment (Kixiron) [inline src/operator/aggregate/fold.rs:71]:
```suggestion
            cursor.map_times(|_t, w| weight.add_assign_by_ref(w));
```

File: src/operator/aggregate/fold.rs [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+            cursor.step_key();
+        }
+
+        // Aggregator must return None iff the input cursor is empty (all keys have
+        // weight 0).
+        if non_empty {
+            Some((self.output)(acc))
+        } else {
+            None
+        }
```
Comment (Kixiron) [inline src/operator/aggregate/fold.rs:86]:
```suggestion
        non_empty.then(|| (self.output)(acc))
```

File: src/operator/aggregate/max.rs [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+        C: Cursor<'s, V, (), T, R>,
+    {
+        let mut result = None;
+
+        while cursor.key_valid() {
+            let mut weight = R::zero();
+
+            cursor.map_times(|_t, w| {
+                weight.add_assign_by_ref(w);
+            });
```
Comment (Kixiron) [inline src/operator/aggregate/max.rs:27]:
```suggestion
            cursor.map_times(|_t, w| weight.add_assign_by_ref(w));
```

File: src/operator/aggregate/min.rs [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+    fn aggregate<'s, C>(&mut self, cursor: &mut C) -> Option<Self::Output>
+    where
+        C: Cursor<'s, V, (), T, R>,
+    {
+        while cursor.key_valid() {
+            let mut weight = R::zero();
+
+            cursor.map_times(|_t, w| {
+                weight.add_assign_by_ref(w);
+            });
```
Comment (Kixiron) [inline src/operator/aggregate/min.rs:27]:
```suggestion
            cursor.map_times(|_t, w| weight.add_assign_by_ref(w));
```

File: src/operator/aggregate/mod.rs [hunk 13]
```diff
@@ -0,0 +1,1057 @@
...
+            DelayedTraceId::new(output_trace.origin_node_id().clone()),
+            output_trace_delayed,
+        );
+        circuit.cache_insert(
+            ExportId::new(output_trace.origin_node_id().clone()),
+            output_trace_export,
+        );
+        circuit.cache_insert(TraceId::new(output.origin_node_id().clone()), output_trace);
+
+        output
```
Comment (Kixiron) [inline src/operator/aggregate/mod.rs:231]: Shouldn't this all be within a cache check?

File: src/operator/aggregate/mod.rs [hunk 14]
```diff
@@ -0,0 +1,1057 @@
...
+    A: Aggregator<Z::Val, (), Z::R> + 'static,
+    O: Clone + IndexedZSet<Key = Z::Key, Val = A::Output> + 'static,
+    O::R: ZRingValue,
+{
+    fn eval(&mut self, i: &Z) -> O {
+        let mut builder = O::Builder::with_capacity((), i.len());
+
+        let mut cursor = i.cursor();
+
+        while cursor.key_valid() {
```
Comment (Kixiron) [inline src/operator/aggregate/mod.rs:363]:
```suggestion
        let mut cursor = i.cursor();
        while cursor.key_valid() {
```

File: src/operator/aggregate/mod.rs [hunk 15]
```diff
@@ -0,0 +1,1057 @@
...
+            if let Some(agg) = self
+                .aggregator
+                .aggregate(&mut CursorGroup::new(&mut cursor, ()))
+            {
+                builder.push((O::item_from(cursor.key().clone(), agg), O::R::one()));
```
Comment (Kixiron) [inline src/operator/aggregate/mod.rs:368]: Kinda don't like that we're forced to clone the key value for this

File: src/operator/aggregate/mod.rs [hunk 16]
```diff
@@ -0,0 +1,1057 @@
...
+                    //     output_cursor.val(),
+                    //     weight
+                    // );
+                    self.output_delta
+                        .push((output_cursor.val().clone(), weight.neg()));
+                }
+                output_cursor.step_val();
+            }
+        }
+        consolidate(&mut self.output_delta);
```
Comment (Kixiron) [inline src/operator/aggregate/mod.rs:619]:
```suggestion
        }

        consolidate(&mut self.output_delta);
```

File: src/trace/cursor/cursor_group.rs [hunk 17]
```diff
@@ -0,0 +1,104 @@
+use crate::{trace::cursor::Cursor, Timestamp};
+use std::marker::PhantomData;
+
+/// A `CursorGroup` iterates over values associated with a single key of a base
+/// cursor of type `C: Cursor<K, V, R, T>`.
+pub struct CursorGroup<'c, K, T, C> {
```
Comment (Kixiron) [inline src/trace/cursor/cursor_group.rs:6]: Lots of missing tests here

**Review by Kixiron — CHANGES_REQUESTED** [review]

File: src/operator/aggregate/mod.rs (hunk 15 above)
Comment (ryzhyk) [reply src/operator/aggregate/mod.rs:368]: Hmm, I really don't see how to avoid this.

Comment (Kixiron) [reply src/operator/aggregate/mod.rs:367]: Yah, it was more of a general gripe

**Comment by lalithsuresh** [comment]
Not sure which aggregates @mbudiu-vmw is implementing for SQL, but the ones that came up with ddlog-dcm were:
* Top-K (ORDER BY + LIMIT)
* min/max
* sum/count
* all/any
* array_agg, set_agg (we used array_agg because we didn't have set_agg, but set_agg would have an efficient incremental implementation unlike array_agg)
* window operators (SQL partition - over)

File: src/trace/cursor/cursor_group.rs (hunk 17 above)
Comment (ryzhyk) [reply src/trace/cursor/cursor_group.rs:6]: @gz kindly promised to share his proptest-based test harness for cursors, which should improve the situation.

**Pull request merged**